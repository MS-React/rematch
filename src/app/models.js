export const game = {
  state: {
    score: 0,
    success: 0,
    fails: 0,
    maxFailures: 5,
    totalToWin: 10,
    timePerProof: 30,
    playing: false,
    player: {
      name: null
    }
  },
  reducers: {
    incrementScore(state, payload) {
      return state.score + payload;
    },
    incrementSuccess(state) {
      state.success + 1;
    },
    incrementFails(state) {
      state.fails + 1;
    },
    proofsLeft(state) {
      state.totalToWin - 1;
    },
    setPlayer(state, payload) {
      state.name = payload;
    },
    pause(state) {
      state.playing = !state.playing;
    }
  }
};
