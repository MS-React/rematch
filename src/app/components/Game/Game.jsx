import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Timer from './Timer';
import PlayerInfo from '../Player/Info';
import ScoreBoard from './Score/Board';
import ScoreSummary from './Score/Summary';
import GameActions from './Actions';
import Proof from './Proof';
import HappyFace from '../HappyFace';

import { createProof } from '../../utils/proofGenerator';

class Game extends React.PureComponent {
  static propTypes = {
    game: PropTypes.shape({}).isRequired,
    player: PropTypes.shape({}).isRequired,
    startGame: PropTypes.func.isRequired,
    resumeAndPause: PropTypes.func.isRequired,
    failure: PropTypes.func.isRequired,
    success: PropTypes.func.isRequired,
    incrementScore: PropTypes.func.isRequired,
    proofsLeft: PropTypes.func.isRequired,
    resetGameState: PropTypes.func.isRequired,
    happyFace: PropTypes.shape({}),
  }

  static defaultProps = {
    happyFace: {},
  }

  state = {
    proof: {
      result: null,
      equation: null,
    },
    userInputResult: null,
    resetTimer: false,
    endGame: false,
  }

  componentWillMount() {
    this.createProof();
  }

  componentDidUpdate() {
    if (this.state.resetTimer === true) {
      this.setState({ // eslint-disable-line
        resetTimer: false,
      });
    }
  }

  onSetResult = (event) => {
    this.setState({
      userInputResult: Number(event.target.value),
    });
  }

  onConfirmProof = () => {
    const { success, failure, incrementScore } = this.props;

    if (this.state.proof.result === this.state.userInputResult) {
      success();
      incrementScore(10);
    } else {
      failure();
    }

    this.nextProof(true);
  }

  onTimeEnd = () => {
    this.props.failure();
    this.nextProof();
  };

  onResumeAndPause = () => {
    this.props.resumeAndPause();
  }

  onStartGame = () => {
    this.props.startGame();
  }

  endGame = () => {
    this.setState({
      endGame: true,
    });
  }

  resetGame = () => {
    this.setState({
      endGame: false,
    }, () => {
      this.props.resetGameState();
    });
  }

  nextProof = (confirmProof = false) => {
    const { game, proofsLeft } = this.props;
    proofsLeft();

    if (game.totalProofs === 0) {
      this.endGame();
    } else {
      this.createProof(confirmProof);
    }
  }

  createProof = (resetTimer = false) => {
    this.setState({
      proof: createProof(),
      resetTimer,
    });
  }

  render() {
    const { game, player, happyFace } = this.props;
    const { resetTimer, proof } = this.state;

    if (!game.started) {
      return (
        <button type="submit" onClick={this.onStartGame}>
          Play!
        </button>
      );
    }

    if (this.state.endGame) {
      return (
        <ScoreBoard
          score={game.score}
          success={game.success}
          fails={game.fails}
          resetGame={this.resetGame}
        />
      );
    }

    return (
      <div className="game">
        <div className="game-info">
          <PlayerInfo player={player} />
          <Timer
            totalTimer={game.timePerProof}
            timeEnd={this.onTimeEnd}
            resetTimer={resetTimer}
            pause={!game.playing}
          />
          <ScoreSummary results={game} />
        </div>
        <Proof proof={proof} />
        <GameActions
          setResult={this.onSetResult}
          confirmProof={this.onConfirmProof}
          resumePause={this.onResumeAndPause}
          playing={game.playing}
        />
        <HappyFace {...happyFace} />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  game: state.game,
  player: state.game.player,
  happyFace: state.happyFace,
});

const mapDispatchToProps = dispatch => ({
  startGame: () => dispatch.game.startGame(),
  success: () => dispatch.game.incrementSuccess(),
  failure: () => dispatch.game.incrementFails(),
  proofsLeft: () => dispatch.game.proofsLeft(),
  incrementScore: score => dispatch.game.incrementScore(score),
  resetGameState: () => dispatch.game.resetGameState(),
  resumeAndPause: () => dispatch.game.resumeAndPause(),
});

export default connect(mapStateToProps, mapDispatchToProps)(Game);
