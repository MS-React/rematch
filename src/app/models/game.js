export default {
  state: {
    score: 0,
    success: 0,
    fails: 0,
    maxFailures: 5,
    totalProofs: 10,
    timePerProof: 30,
    playing: false,
    player: {
      name: null
    },
    proofs: []
  },
  reducers: {
    incrementScore: (state, payload) => ({
      ...state,
      score: state.score + Number(payload)
    }),
    incrementSuccess: (state) => ({
      ...state,
      success: ++state.success
    }),
    incrementFails: (state) => ({
      ...state,
      fails: ++state.fails
    }),
    proofsLeft: (state) => ({
      ...state,
      totalProofs: --state.totalProofs
    }),
    setPlayer: (state, payload) => {
      return {
        ...state,
        player: {
          name: String(payload)
        }
      }
    },
    resumeAndPause: (state) => ({
      ...state,
      playing: !state.playing
    })
  }
};
