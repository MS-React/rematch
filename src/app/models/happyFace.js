export default {
  state: {
    display: false,
    showHappyFace: false,
    showUnhappyFace: false,
  },
  reducers: {
    showHappyFace: state => ({
      ...state,
      display: true,
      showHappyFace: true,
      showUnhappyFace: false,
    }),
    showUnhappyFace: state => ({
      ...state,
      display: true,
      showUnhappyFace: true,
      showHappyFace: false,
    }),
  },
};
