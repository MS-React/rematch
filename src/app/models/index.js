export { default as game } from './game';
export { default as happyFace } from './happyFace';
