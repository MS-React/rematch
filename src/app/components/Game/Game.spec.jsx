import React from 'react';
import { shallow } from 'enzyme';
import { init } from '@rematch/core';
import renderer from 'react-test-renderer';
import * as models from '../../models';

import Game from './Game';

jest.mock('../../utils/proofGenerator', () => ({
  createProof: jest.fn().mockReturnValue({
    equation: '1 - 1',
    result: 0,
  }),
}));

describe('<Game />', () => {
  let store;
  let wrapper;

  beforeAll(() => {
    store = init({ models });
    store.getState().game.player = {
      name: 'John Doe',
    };
    store.dispatch.game = {
      ...store.dispatch.game,
      startGame: jest.fn(),
      resumeAndPause: jest.fn(),
      incrementFails: jest.fn(),
      incrementSuccess: jest.fn(),
      incrementScore: jest.fn(),
      proofsLeft: jest.fn(),
      resetGameState: jest.fn(),
    };
  });

  describe('initializes', () => {
    beforeEach(() => {
      wrapper = shallow(<Game store={store} />);
    });

    it('should render a button to start', () => {
      expect(renderer.create(wrapper.html()).toJSON())
        .toMatchSnapshot();
    });

    describe('start', () => {
      it('should dispatch startGame and change the state from store', () => {
        wrapper.dive().find('button').simulate('click');
        expect(store.dispatch.game.startGame).toHaveBeenCalled();
      });

      it('should render <PlayerInfo> , <Timer>, <ScoreSummary>, <Proof> and <GameActions> components', () => {
        store.getState().game.started = true;
        expect(renderer.create(wrapper.html()).toJSON())
          .toMatchSnapshot();
      });
    });
  });

  describe('is playing', () => {
    let gameActionsComponent;
    let timerComponent;

    beforeEach(() => {
      store.getState().game.started = true;
      wrapper = shallow(<Game store={store} />).dive();
      gameActionsComponent = wrapper.find('GameActions');
      timerComponent = wrapper.find('Timer');
    });

    describe('confirmProof', () => {
      describe('success', () => {
        beforeEach(() => {
          gameActionsComponent.prop('setResult')({ target: { value: 0 } });
          gameActionsComponent.prop('confirmProof')();
        });

        it('should dispatch incrementSuccess callback from store', () => {
          expect(store.dispatch.game.incrementSuccess).toHaveBeenCalled();
        });

        it('should dispatch incrementScore callback from store', () => {
          expect(store.dispatch.game.incrementScore).toHaveBeenCalled();
        });
      });

      describe('fails', () => {
        beforeEach(() => {
          gameActionsComponent.prop('setResult')({ target: { value: -1 } });
          gameActionsComponent.prop('confirmProof')();
        });
        it('should dispatch incrementFails callback from store', () => {
          expect(store.dispatch.game.incrementFails).toHaveBeenCalled();
        });
      });
    });

    describe('pause', () => {
      beforeEach(() => {
        gameActionsComponent.prop('resumePause')({ target: { value: -1 } });
      });
      it('should dispatch resumeAndPause callback from store', () => {
        expect(store.dispatch.game.resumeAndPause).toHaveBeenCalled();
      });
    });

    describe('timer ends', () => {
      beforeEach(() => {
        store.getState().game.totalProofs = 0;
        timerComponent.prop('timeEnd')();
      });

      it('should dispatch incrementFails callback from store', () => {
        expect(store.dispatch.game.incrementFails).toHaveBeenCalled();
      });
    });
  });

  describe('finish', () => {
    let scoreBoardComponent;
    beforeEach(() => {
      store.getState().game.started = true;
      wrapper = shallow(<Game store={store} />).dive();
      wrapper.setState({
        endGame: true,
      });
      scoreBoardComponent = wrapper.find('ScoreBoard');
    });

    it('should display the <ScoreBoard />', () => {
      expect(scoreBoardComponent).toHaveLength(1);
    });

    describe('reset game', () => {
      beforeEach(() => {
        scoreBoardComponent.prop('resetGame')();
      });

      it('should dispatch resetGameState callback from store', () => {
        expect(store.dispatch.game.resetGameState).toHaveBeenCalled();
      });
    });
  });
});
