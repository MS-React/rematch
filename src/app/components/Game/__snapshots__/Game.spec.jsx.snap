// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> initializes should render a button to start 1`] = `"<button type=\\"submit\\">Play!</button>"`;

exports[`<Game /> initializes start should render <PlayerInfo> , <Timer>, <ScoreSummary>, <Proof> and <GameActions> components 1`] = `"<div class=\\"game\\"><div class=\\"game-info\\"><div class=\\"playerInfo\\">Hello! John Doe</div><div class=\\"timer\\">30</div><div class=\\"score-summary\\">Total Score: 0 <br/>Success: 0 <br/>Failures: 0<br/></div></div><div class=\\"game-proof\\">1 - 1 = ?</div><div class=\\"game-actions\\"><input type=\\"text\\"/><button type=\\"submit\\" class=\\"action-confirmProof\\">Confirm</button><button type=\\"submit\\">Resume</button></div><div class=\\"happyFace\\"></div></div>"`;
